src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs

```diff
@@ -286,12 +287,41 @@ unsafe long GetFileLengthCore()
             {
                 Interop.Kernel32.FILE_STANDARD_INFO info;
 
-                if (!Interop.Kernel32.GetFileInformationByHandleEx(this, Interop.Kernel32.FileStandardInfo, &info, (uint)sizeof(Interop.Kernel32.FILE_STANDARD_INFO)))
+                if (Interop.Kernel32.GetFileInformationByHandleEx(this, Interop.Kernel32.FileStandardInfo, &info, (uint)sizeof(Interop.Kernel32.FILE_STANDARD_INFO)))
+                {
+                    return info.EndOfFile;
+                }
+
+                if (_path is null || !PathInternal.IsDevice(_path))
                 {
                     throw Win32Marshal.GetExceptionForLastWin32Error(Path);
                 }
 
-                return info.EndOfFile;
+                // In theory GetFileInformationByHandleEx might fail, then DeviceIoControl succeed (last error set to ERROR_SUCCESS)
+                // but return fewer bytes than requested. The error is stored and in such case exception for the first failure is going to be thrown.
+                int lastError = Marshal.GetLastWin32Error();
```

Comment:
I would move this line before `if (_path is null || !PathInternal.IsDevice(_path))` and also use it to throw the exception when `(_path is null || !PathInternal.IsDevice(_path))`. 

`PathInternal.IsDevice` can modify the last error. It won't do it today, but it is fragile to depend on invariants like that.